django/db/models/sql/query.py

```diff
@@ -1030,19 +1032,22 @@ def resolve_expression(self, query, *args, **kwargs):
         for key, value in clone.annotations.items():
             resolved = value.resolve_expression(query, *args, **kwargs)
             if hasattr(resolved, 'external_aliases'):
-                resolved.external_aliases.update(clone.alias_map)
+                resolved.external_aliases.update(clone.external_aliases)
             clone.annotations[key] = resolved
         # Outer query's aliases are considered external.
-        clone.external_aliases.update(
-            alias for alias, table in query.alias_map.items()
-            if (
-                isinstance(table, Join) and table.join_field.related_model._meta.db_table != alias
-            ) or (
-                isinstance(table, BaseTable) and table.table_name != table.table_alias
+        for alias, table in query.alias_map.items():
+            clone.external_aliases[alias] = (
```

Comment:
I guess it would be easier to understand, if you'd name the separate boolean expression first before combining them, eg:
```python
is_join = isinstance(table, Join)
is_base_table = isinstance(table, BaseTable)
clone.external_aliases[alias] = is_join and not is_base_table
```

Something along those lines.

Reply:
The current form is fine, IMO. I don't think that
```python
is_join = isinstance(table, Join)
is_base_table = isinstance(table, BaseTable)
clone.external_aliases[alias] = (
    (is_join and table.join_field.related_model._meta.db_table != alias) or
    (is_base_table and table.table_name != table.table_alias)
)
```
is more readable.